### ISSUE TITLE
Incompatibility or tenant connection scoping with ASGI servers

### ISSUE THREAD

**Issue by ey3ball** [django-tenants/django-tenants#820]
This initially came up while trying to setup django-tenants in SubFolder mode with an existing project that also uses django channels.

Accessing a tenant url would lead to the following error while processing any request :
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/asgiref/sync.py", line 472, in thread_handler
    raise exc_info[1]
  File "/usr/local/lib/python3.9/site-packages/django/core/handlers/exception.py", line 42, in inner
    response = await get_response(request)
  File "/usr/local/lib/python3.9/site-packages/django/core/handlers/base.py", line 235, in _get_response_async
    callback, callback_args, callback_kwargs = self.resolve_request(request)
  File "/usr/local/lib/python3.9/site-packages/django/core/handlers/base.py", line 313, in resolve_request
    resolver_match = resolver.resolve(request.path_info)
  File "/usr/local/lib/python3.9/site-packages/django/urls/resolvers.py", line 651, in resolve
    sub_match = pattern.resolve(new_path)
  File "/usr/local/lib/python3.9/site-packages/django/urls/resolvers.py", line 646, in resolve
    match = self.pattern.match(path)
  File "/code/examples/tenant_subfolder_tutorial/../../django_tenants/urlresolvers.py", line 50, in match
    tenant_prefix = self.tenant_prefix
  File "/code/examples/tenant_subfolder_tutorial/../../django_tenants/urlresolvers.py", line 33, in tenant_prefix
    domain=connection.tenant.domain_subfolder,

Exception Type: AttributeError at /clients/test/
Exception Value: 'FakeTenant' object has no attribute 'domain_subfolder'
```

I was able to track this down and reproduce the issue with the example `tenant_subfolder_tutorial`. As soon as the project behind an ASGI server, the previous error occurs when accessing a subfolder tenant.

Steps to reproduce : 
* Setup the project and subfolder tenant
  * Setup the tenant_subfolder_tutorial example project and run it
  * Create one public tenant, then one `test` tenant, with domain = `test`
  * Head over to http://localhost:8088/clients/test/ and verify the project is properly operating
![image](https://user-images.githubusercontent.com/96346/186680437-6357fe09-71db-472f-bdd9-4aa0a88cea35.png)
* Enable ASGI trigger the issue
  * Add uvicorn in requirements and create an asgi.py file (or apply https://github.com/ey3ball/django-tenants/commit/e5cc6fa57ed9c35cd5d3cad104d0c6dc90275031)
  * Start project through asgi server with `gunicorn tenant_subfolder_tutorial.asgi:application --bind 0.0.0.0:8088 -k uvicorn.workers.UvicornWorker` (or apply https://github.com/django-tenants/django-tenants/commit/31ed2eb81590b2d0c4327afa53e0cbfc94ccda45 and run it with compose)
  * Attempt to access http://localhost:8088/clients/test/
  * Request fails with FakeTenant backtrace above

![image](https://user-images.githubusercontent.com/96346/186682614-39770f58-a7eb-4e4f-99d2-b1ff64d10c8a.png)

Preliminary analysis : 

* in [django_tenants/middleware/subfolder.py:](https://github.com/django-tenants/django-tenants/blob/master/django_tenants/middleware/subfolder.py#L72) the connection object is used to set a tenant
* It is later accessed by https://github.com/ey3ball/django-tenants/blob/master/django_tenants/urlresolvers.py#L33
* **However at that point instead of getting the connection information we though we saved in the first step, we get a default FakeTenant object that defaults to the public tenant** (https://github.com/ey3ball/django-tenants/blob/master/django_tenants/postgresql_backend/base.py#L187)
* Because domain_subfolder is not set on this object a backtrace ensues

At this moment I think this boils down to the fact that we can't rely on the connection object staying the same during the whole processing of the request. In WSGI mode I guess we can rely on 1 request = 1 connection because there will be one thread dedicated to servicing each request. In ASGI mode it seems this does not hold anymore.

Although this came up with the SubFolder Middleware, this might be a more general issue of pinning data on the Connection object in django_tenants middlewares

**Comment by ey3ball** [comment]
Digging a bit deeper into django's ASGI code, it seems the ASGI framework will make every effort to guarantee that sync functions al execute under the same thread/context. Middleware methods are wrapped using SyncToAsync (https://github.com/django/asgiref/blob/6a0cae060c0d96d7bf0d5a3dc38dbde090ab3e1a/asgiref/sync.py#L295) which makes every effort to guarantee that.

However for the SubFolder middleware, we also hook into the urlconf which ultimately wraps the offending function : https://github.com/django-tenants/django-tenants/blob/master/django_tenants/urlresolvers.py#L27

In Django's ASGI code the request processing code in the async case is here : 
* https://github.com/django/django/blob/71902e0d9f93670c4f93ff9d66095b0e571be74b/django/core/handlers/base.py#L162

Where _middleware_chain is a stack of middleware calls adapted through sync_to_async as required.

The tail of the middleware chain is the request processing itself wrapped by `_get_response_async`

Looking at the code of _get_response_async https://github.com/django/django/blob/71902e0d9f93670c4f93ff9d66095b0e571be74b/django/core/handlers/base.py#L228 we can see that it begins with a call to `self.resolve_request(request)`, this calls ultimately leads to `TenantPrefixPattern.tenant_prefix()`

One thing to note, is that contrary to the middleware objects, the call is not wrapped by sync/async gates. So I believe this means  tenant_prefix it called from the async world which would explain why the connection object does not contain the expected tenant information.

TLDR:
* In asgi mode tenant_prefix is not called in the same context at the rest of the request processing code
* This is specific to the dynamic url resolution which is done in subfolder mode and does not affect the regular MiddleWare

**Comment by ey3ball** [comment]
👋 @lorinkoz (initial contributor of the subfolder code) in case you have some insights

I'm trying to find the right way to fix this, currently the domain object is looked up during the path matching due to tenant_prefix. Which is the root cause of the crash.

I'm thinking that since the urlconf is set during process_request in TenantSubfolderMiddleware, creating tenant specific urlconfs and setting them at this point would fix the issue.

However it looks like this is something you wanted to avoid considering we have a single urlconf with dynamic resolution at the moment.

**Comment by ey3ball** [comment]
I've tried the approach above (WIP patch here : https://github.com/ey3ball/django-tenants/commit/b995f77f1b4f0a4d77f945a913aecbe1a58a6de3), this seem to confirm what I've discovered so far, with this patch I can successfully acces the homepage of my tenant in subfolder mode with an asgi frontend.

Some more test required, I guess my approach might nos support reverse() at the moment since I'm actually registering the generated urlconf module anywhere

**Comment by lorinkoz** [comment]
I don't have any possible solution in mind right now. I would look into it hopefully over the weekend.

I wanted to experiment with storing tenant information outside of the connection (e.g. asgiref.local) but ran into other problems before.

Godspeed with your efforts.

**Comment by lorinkoz** [comment]
> However it looks like this is something you wanted to avoid considering we have a single urlconf with dynamic resolution at the moment.

Yes, ideally the subfolder approach should work without requiring changes in the project urlconf.

**Comment by ey3ball** [comment]
Thanks for the feedback ! I'll keep that in mind too, see if I can think of something.

In the meantime I'll clean up my initial approach so that we at least have a working fix.

**Comment by lorinkoz** [comment]
Storing the active tenant in `asgiref.local` instead of the connection object seems promising (it's currently working for me [here](https://github.com/lorinkoz/django-pgschemas/pull/120) but I need to test more). For context, Django itself uses `asgiref.local` when storing globals like active timezone and active language.

The idea for the `DatabaseWrapper` is to "remember" the last search path set, and decide if it needs to set a new search path based on comparing the search path of the active tenant vs the last one set. This should more less map to the current behavior.

What do you think?

**Comment by lorinkoz** [comment]
Alternatively.- would it be too crazy to try to import the connection inside `tenant_prefix `?

**Comment by ey3ball** [comment]
This seems promising indeed, although I'm not 100% of my understanding of asgiref.local at the moment.

I can see that it is used already by django's connection manager : https://github.com/django/django/blob/main/django/utils/connection.py#L41, however the actual handler uses thread_critical here : https://github.com/django/django/blob/master/django/db/utils.py#L145 (while your implementation does not since thread_critical defaults to False in Local https://github.com/django/asgiref/blob/main/asgiref/local.py#L33)

From what I gather from the documentation there your usage looks correct though

**Comment by ey3ball** [comment]
Per the comment above I don't think importing connection into tenant_prefix would work because of the Local(thread_critical=True), asgiref,local seems cleaner since we have control over the sharing flags instead of relying on whatever magic the connection class performs

**Comment by awachat** [comment]
@ey3ball @lorinkoz, our team is also facing the same issue. Any solutions for now?

**Comment by ey3ball** [comment]
The solution is either to transition to django-pgschemas which has a modernized approach with this issue fixed. With django-tenants as far as I know the only solution is to pull my fixes here : https://github.com/django-tenants/django-tenants/pull/822

I haven't heard back about integrating it so far. I can rebase on the latest master and have another go at it

**Comment by lorinkoz** [comment]
@tomturner and everyone, sorry that I haven't had the time to follow through this issue. I can confirm that storing the current tenant outside of the Django DB connection works, but it takes a major refactoring. @ey3ball Would you please include some tests in your PR to ensure the fixes work as expected?

**Comment by tomturner** [comment]
@ey3ball I agree with @lorinkoz it needs some tests

**Comment by craigRSA** [comment]
Hi,
What is the latest here? We are facing the same issue and would love a fix

**Comment by BigVaibhav** [comment]
We moved to https://django-pgschemas.readthedocs.io/. It's in production and works great for us.
